the-collab-lab/tcl-6-smart-shopping-list
# Sb pja listshare

## Type of Changes



|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|    | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

## Updates

### Before


![Screen Shot 2020-04-14 at 1 33 01 PM](https://user-images.githubusercontent.com/7929227/79927083-774c4480-83f3-11ea-86c1-c17a21639471.png)


### After


![Screen Shot 2020-04-21 at 5 14 48 PM](https://user-images.githubusercontent.com/7929227/79927133-a5ca1f80-83f3-11ea-8f43-e0a56208ad64.png)

![Screen Shot 2020-04-21 at 5 15 55 PM](https://user-images.githubusercontent.com/7929227/79927179-cb572900-83f3-11ea-867c-02141a267e86.png)

![Screen Shot 2020-04-22 at 1 52 28 PM](https://user-images.githubusercontent.com/7929227/80032801-8b9a4b00-84a0-11ea-9aa7-d74f6b2f6ccd.png)

![Screen Shot 2020-04-22 at 1 58 39 PM](https://user-images.githubusercontent.com/7929227/80033349-678b3980-84a1-11ea-8c40-366af35b0fda.png)


## Related Issue


This PR closes #7 

## Description


### Objective 
Tokens can be shared with other users to allow them to co-manage a given list.

Joining an existing list consists of the following:

- User enters an existing token
- Save the token to localStorage
- Show the user the existing list

### Steps to accomplish to build the feature
1. User enters an existing token: 

- [x]  create form field to enter token 
- [x]   save user input to state 
- [x]   verify that state has been set to user input 

2. User submits the token: 
- [x] query the db for items in the `items` collection with the given token (based on current data structure)

3. IF the token is valid (the query results length is not equal to 0):

- [x]  Save the token to localStorage 
- [x]  Show the user the existing list

4. IF the token is NOT valid (the query results length is equal to 0): 

- [x]   show an error if the token does not exist 
- [x]   allow them to try again or 
- [x]   allow them to create a new list 

## Acceptance Criteria

AC:

1. When the user does not already have a token in localStorage, on the onboarding/home screen, the user can enter an existing token
2. Entering the token and hitting submit saves the token to localStorage, effectively giving them joint control of the list
3. On submit, show an error if the token does not exist
4. If they get an error message, allow them to try again or create a new list

## Testing Steps / QA Criteria
1)  Git pull the branch using `git pull origin sb-pja-listshare` and `git checkout sb-pja-listshare` to switch to the branch.
2) Run `npm install` to add dependencies, then `npm start` to run the app.

TO TEST: 
1. You might need to clear your localStorage of `user_token` -> run `localStorage.removeItem('user_token')` in your browser console. Refresh the page.
2. `wile cain woo` is a valid user token.

- Given: A `user_token` exists in `localStorage` the app will show the shopping list component.
- Given: A `user_token` does not exist in `localStorage`, the app will present the user with the choice to create a token **or** search for and use a `user_token` given to them by another user. 
- Given: An invalid shared token, a warning message and the options of submitting a shared token **or** creating a new token are still available.
- Given: A valid shared token should present the user with the expected shopping list.
